lsp-mode.el

```diff
@@ -4115,35 +4161,31 @@ PLIST is the additional data to attach to each candidate."
           ((and lsp--capf-cache
                 (s-prefix? (car lsp--capf-cache)
                            (buffer-substring-no-properties bounds-start (point))))
-           (-let [(&plist :start-point) (cddr lsp--capf-cache)]
-             (unless (and start-point (> start-point (point-max)))
-               (apply #'lsp--capf-filter-candidates
-                      pred
-                      (buffer-substring-no-properties (or start-point (point)) (point))
-                      (cdr lsp--capf-cache)))))
+           (apply #'lsp--capf-filter-candidates (cdr lsp--capf-cache)))
           (t
            (-let* ((resp (lsp-request-while-no-input "textDocument/completion"
                                                      (plist-put (lsp--text-document-position-params)
                                                                 :context (ht ("triggerKind" 1)))))
-                   (items (lsp--sort-completions (cond
-                                                  ((seqp resp) resp)
-                                                  ((hash-table-p resp) (gethash "items" resp)))))
-                   (start (or (car (-some-> (lsp-elt items 0)
-                                     (lsp--ht-get "textEdit" "range")
-                                     lsp--range-to-region))
-                              bounds-start))
-                   (prefix (buffer-substring-no-properties start (point)))
+                   (items (->> (lsp--sort-completions (cond
+                                                       ((seqp resp) resp)
+                                                       ((hash-table-p resp) (gethash "items" resp))))
+                               (-map (lambda (item)
+                                       (puthash "emacsStartPoint_"
+                                                (lsp--capf-guess-prefix item bounds-start)
+                                                item)
+                                       item))))
                    (prefix-line (buffer-substring-no-properties (point-at-bol) (point))))
              (setf done? (or (seqp resp)
                              (not (gethash "isIncomplete" resp)))
                    lsp--capf-cache (when (and done? (not (seq-empty-p items)))
                                      (list (buffer-substring-no-properties bounds-start (point))
-                                           items
-                                           :start-point start
+                                           (lsp--capf-make-cache items)
+                                           nil
```

Comment:
I convert the cache to PLIST, the nil is actually the original completion items returned from language server. In case we already has the cache, it's not used, so passing nil here